lib/Dialect/FIRRTL/Transforms/LowerBundleVectorTypes.cpp

```diff
@@ -0,0 +1,548 @@
+      loweredAttrs.push_back(opAttr);
+      continue;
+    }
+    auto targetAttr = di.get("target");
+    if (!targetAttr) {
+      loweredAttrs.push_back(opAttr);
+      continue;
+    }
+
+    ArrayAttr subFieldTarget = targetAttr.cast<ArrayAttr>();
+    SmallString<16> targetStr;
+    for (auto fName : subFieldTarget) {
+      std::string fNameStr = fName.cast<StringAttr>().getValue().str();
+      // The fNameStr will begin with either '[' or '.', replace it with an
+      // '_' to construct the suffix.
+      fNameStr[0] = '_';
+      // If it ends with ']', then just remove it.
+      if (fNameStr.back() == ']')
+        fNameStr.erase(fNameStr.size() - 1);
+
+      targetStr += fNameStr;
+    }
+    // If no subfield attribute, then copy the annotation.
+    if (targetStr.empty()) {
+      loweredAttrs.push_back(opAttr);
+      continue;
+    }
+    // If the subfield suffix doesn't match, then ignore the annotation.
+    if (suffix.find(targetStr.str().str()) != 0)
+      continue;
+
+    NamedAttrList modAttr;
+    for (auto attr : di.getValue()) {
+      // Ignore the actual target annotation, but copy the rest of annotations.
+      if (attr.first.str() == "target")
+        continue;
+      modAttr.push_back(attr);
+    }
+    loweredAttrs.push_back(
+        DictionaryAttr::get(annotations.getContext(), modAttr));
+  }
+}
+
+/// Copy annotations from \p annotations into a new AnnotationSet and return it.
+/// This removes annotations with "target" key that does not match the field
+/// suffix.
+static AnnotationSet filterAnnotations(AnnotationSet annotations,
+                                       StringRef suffix) {
+  if (annotations.empty())
+    return annotations;
+  SmallVector<Attribute> loweredAttrs;
+  filterAnnotations(annotations.getArrayAttr(), loweredAttrs, suffix);
+  return AnnotationSet(ArrayAttr::get(annotations.getContext(), loweredAttrs));
+}
+
+//===----------------------------------------------------------------------===//
+// Module Type Lowering
+//===----------------------------------------------------------------------===//
+
+namespace {
+using LoweredValues = SmallVector<std::pair<Value, bool>>;
+class DoLowering {
+public:
+  DoLowering(FModuleOp &m, ImplicitLocOpBuilder *b, MLIRContext *c)
+      : module(m), builder(*b), context(*c) {
+
+    auto *body = module.getBodyBlock();
+    SmallVector<BlockArgument, 8> args(body->args_begin(), body->args_end());
+    // Initialize the worklist with block arguments.
+    for (BlockArgument arg : args)
+      if (arg.getType().isa<FIRRTLType>())
+        worklist.push_back(arg);
+
+    runLowering();
+    // Update the module with the new arguments.
+    cleanupModule();
+  }
+
+private:
+  SmallVector<Value> worklist;
+  FModuleOp &module;
+  ImplicitLocOpBuilder &builder;
+  MLIRContext &context;
+  DenseMap<Value, LoweredValues> loweredValueMap;
+  // State to track the new attributes for the module.
+  SmallVector<NamedAttribute, 8> newModuleAttrs;
```

Comment:
Isn't there a ModulePortInfo which contains these several things you are storing broken out in a vector?